---
#############################
#  Input Module Parameters  #
#############################
input:
  global_frame_name: $(arg odom_frame)
  sensor_frame_name: $(arg sensor_frame)
  sensor:
    type: camera
    width: 1200
    height: 680
    cx: 599.5
    cy: 339.5
    fx: 600.0
    fy: 600.0
    min_range: $(arg sensor_min_range)
    max_range: $(arg sensor_max_range)
    extrinsics:
      type: identity
  subscriber:
    type: $(arg subscriber_type)
    input_images_are_const: false
    color_image_topic: color/image_raw
    depth_image_topic: depth/image_rect
    semantic_image_topic: semantic/image_raw
    queue_size: 100
#############################################
#  Active-Window Reconstruction Parameters  #
#############################################
reconstruction:
  use_visualizer: true
  active_window_viz:
    global_frame_name: $(arg odom_frame)
    color_mode: "color"
    slice_height: -0.5
    slice_height_is_relative: true
    activity_shading_intensity: 0.6
    object_color_mode: 0
    dynamic_point_scale: 0.1
  active_window:  # taken from Khronos
    verbosity: 1
    find_visible_blocks_accurately: false
    max_frame_data_buffer_size: 300
    store_every_n_frame_data: 1
    min_object_allocation_confidence: 0.5
    min_object_volume: 0.0  # m^3
    min_object_reconstruction_confidence: 0.5
    scale_object_reconstruction_confidence_with_weight: true
    min_object_reconstruction_observations: 0
    object_reconstruction_resolution: -0.02  # Positive: voxel size in meters. Negative: fraction of the extent. 0: Skip. (-0.02)
    volumetric_map:
      voxel_size: 0.12  # m
      truncation_distance: 0.2  # m (Usually 2-3x voxel size)
      voxels_per_side: 16
    object_detector:
      type: $(arg object_detection_type)
      min_cluster_size: 100  # pixels
      use_full_connectivity: true
      use_3d: true
      grid_size: 0.1  # m
      max_range: $(arg sensor_max_range)
    tracker:
      type: max_iou   # 'max_iou'
      track_by: voxels   # 'pixels', 'voxels', 'bounding_box'
      min_semantic_iou: 0.3
      min_cosine_sim: 0.7
      min_cross_iou: 0.1
      temporal_window: $(arg temporal_window_size_s)
      min_num_observations: 3
    projective_integrator:
      max_weight: 100000
      interpolation_method: adaptive   # "nearest", "bilinear", "adaptive"
      verbosity: 1
    tracking_integrator:
      temporal_window: $(arg temporal_window_size_s)
      verbosity: 1
#########################
#  Frontend Parameters  #
#########################
frontend:
  type: LLMFrontend
  min_object_merge_similiarity: 0.8
  pgmo:
    horizon: 60.0
    d_graph_resolution: 2.5
    output_mesh_resolution: 0.02
  min_object_vertices: 0
  prune_mesh_indices: true
  gvd:
    max_distance_m: 4.5
    min_diff_m: 0.1
    min_basis_for_extraction: 1
    voronoi_config:
      min_distance_m: 0.30
      parent_l1_separation: 20
      parent_cos_angle_separation: 0.2
  graph_extractor:
    type: CompressionGraphExtractor
    compression_distance_m: 1.0
    min_node_distance_m: 0.4
    min_edge_distance_m: 0.25
    node_merge_distance_m: 0.5
    add_heuristic_edges: false
########################
#  Backend Parameters  #
########################
backend:
  type: BackendModule
  dsg:
    places_merge_distance_tolerance_m: 0.4
  pgmo:
    run_mode: 0  # kimera_pgmo run mode FULL (required)
    use_msg_time: true
    embed_trajectory_delta_t: 5.0
    num_interp_pts: 3
    interp_horizon: 10.0
    enable_sparsify: false
    trans_node_dist: 1.0
    rot_node_dist: 1.2
    rpgo:
      odom_trans_threshold: 0.05
      odom_rot_threshold: 0.01
      pcm_trans_threshold: -1
      pcm_rot_threshold: -1
      gnc_alpha: 0.9
      gnc_mu_step: 1.6
      gnc_fix_prev_inliers: true
      verbosity: UPDATE
      solver: LM
    covariance:
      odom: 1.0e-02
      loop_close: 5.0e-02
      sg_loop_close: 1.0e-01
      prior: 1.0e-02
      mesh_mesh: 1.0e-02
      pose_mesh: 1.0e-02
      place_mesh: 1.0e-02
      place_edge: 10.0
      place_merge: 10.0
      object_merge: 10.0
  enable_rooms: false
  regions:
    extractor:
      type: PlaceClustering
      min_assocation_iou: 0.8
      clustering:
        type: $(arg places_clustering_type)
        stop_value: 0.0
        tasks:
          type: RosEmbeddingGroup
        metric:
          type: cosine
        merge:
          type: weighted_mean
